Package guidelines in developer docs (#14053)
wagtail/wagtail: docs/packages/index.md

(packages_guide)=

# Packages

Wagtail benefits from a rich ecosystem of packages. Many Python and Django packages work with Wagtail out of the box, and there is a growing collection of Wagtail-specific packages for extending the CMS.

## Finding existing packages

The official [Wagtail third-party packages directory](https://wagtail.org/packages/) lists packages organized by category. Forms, migrations, media, integrations, SEO, and more. [Awesome Wagtail](https://github.com/springload/awesome-wagtail) offers a curated list of packages, articles, and resources.

> Many general-purpose Django or Python packages also work directly, or can be integrated directly inside the CMS with little work. Check out the [Django Packages](https://djangopackages.org/) directory, or search directly on the [Python Package Index (PyPI)](https://pypi.org/). You can use the classifier [`Framework :: Wagtail`](https://pypi.org/search/?q=&o=&c=Framework+%3A%3A+Wagtail) to find packages that explicitly target Wagtail compatibility.

Reviewer: If this were Django docs I think we should drop such wordings, but if this were intentional for better marketing then I don't mind keeping it

```suggestion
Many general-purpose Django or Python packages also work directly, or can be integrated directly inside the CMS. Check out the [Django Packages](https://djangopackages.org/) directory, or search directly on the [Python Package Index (PyPI)](https://pypi.org/). You can use the classifier [`Framework :: Wagtail`](https://pypi.org/search/?q=&o=&c=Framework+%3A%3A+Wagtail) to find packages that explicitly target Wagtail compatibility.
```

### Documentation

This is crucial to the success of your package.

- We recommend writing all documentation in vanilla Markdown ([CommonMark](https://commonmark.org/)), with minimal to no use of platform-specific extensions.
- Start with a README with basic information about all aspects of the package. Include:
    - Support targets for Python, Django, Wagtail.
    - Simple installation guide.
    - Links to all relevant resources about using or working on the package (documentation, support, security, changelog, etc.).
    - A screenshot of the package user interface, if applicable.
- Create separate files for all aspects of the package to be documented. We recommend the [Diátaxis](https://diataxis.fr/) documentation structure.
    - Produce reference documentation for all public APIs of the package.
    - Create how-to material for common usage scenarios.
    - Add a beginner tutorial for first-time users of the package.
    - Add a user guide if the package includes a user interface.
    - Create a `CONTRIBUTING.md` guide for contributors.

> - Consider publishing the package documentation as a website.

Reviewer: Should we recommend or mention RTD here?

Author: I am hesitant to do that without further R&D on current options for documentation site tooling. See https://github.com/wagtail/cookiecutter-wagtail-package/issues/81.

Reviewer: Sounds fair to me!

- Use [@wagtail/stylelint-config-wagtail](https://github.com/wagtail/stylelint-config-wagtail) for CSS linting, and [@wagtail/eslint-config-wagtail](https://github.com/wagtail/eslint-config-wagtail) for JS linting.
- Write unit tests using Django’s built-in test framework.
- Aim for unit test coverage above 90% (excluding migrations).
- Set up continuous integration to automatically test every change to the package. Continuous integration should include:
    - Matrix testing across a representative range of supported Python / Django / Wagtail versions.
    - Security scanning of project code and dependencies.
    - Recommended: [nightly testing of official plugins](https://wagtail.org/nightly-packages/).
- Set up automation to manage the project’s dependencies.

### Releases

- Follow [Semantic Versioning](https://semver.org/) and [PEP 440 – Version Identification and Dependency Specification](https://peps.python.org/pep-0440/).
- Document changes in a `CHANGELOG.md` following the [Keep a Changelog](https://keepachangelog.com/) format.
- Document upgrade considerations when releasing breaking changes to the package.

> - Use a git `vx.y.z` git tag and [repository releases](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository).

Author: ```suggestion
- Use a git `va.b.c` git tag and [repository releases](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository).
```

- Release to PyPI with a [Trusted Publisher](https://docs.pypi.org/trusted-publishers/using-a-publisher/) set up in Continuous Integration.

> - Use a `stable/x.y` branch when working on patch releases to previous versions.

Reviewer: Not sure if it matters but our format is `stable/a.b.x`

```suggestion
- Use a `stable/a.b.x` branch when working on patch releases to previous versions.
```

Though it might be confusing because the `x` is literal.

If we want to apply `a.b.x` here, we might want to update the "Use a git `vx.y.z`" line above to "Use a git `va.b.c`" for consistency.

### Collaborative maintenance

- Follow the [Wagtail Code of Conduct](https://wagtail.org/code-of-conduct/).

> - For End of Life (EOL) packages, add a message to the package README, and visible on PyPI as well. For official packages, we additionally transfer them to the [wagtail-deprecated organization in GitHub](https://github.com/wagtail-deprecated) organisation.

Reviewer: ```suggestion
- For End of Life (EOL) packages, add a message to the package README, and visible on PyPI as well. For official packages, we additionally transfer them to the [wagtail-deprecated organization on GitHub](https://github.com/wagtail-deprecated).
```

> - Share PyPI access to the package with at least one or two other people to reduce the bus factor on maintainers.

Reviewer: Should we link to [bus factor](https://en.wikipedia.org/wiki/Bus_factor)? 😆 

```suggestion
- Share PyPI access to the package with at least one or two other people to reduce the [bus factor](https://en.wikipedia.org/wiki/Bus_factor) on maintainers.
```

Author: good idea!
